### PR TITLE
Add script to export grafana snapshots

### PR DESCRIPTION
**Pull request by richardwxn**
[istio/#10285](https://github.com/istio/istio/issues/10285)
Old revision 
1. Add dashboard and snapshot related api wrapper
2. Schedule job to export graphs to local or to raintank

New Revision change:
1. Add logic to persist snapshot url and dashboard related information to google cloud datastore, these entries can be queried or filter by dashboard name, creation time and tags on google cloud console. Use datastore instead of gcs because it is easier to query
2.  Get rid of using post grafana snapshot api, this api is not friendly to use externally now, it requires a dashboard object as input, however the dashboard object obtained by dashboard api is not fulfilled with panel data. So there is data missing after export to raintank. There is no easy way or guidance now to fill the data and we have to manually query datasource and populate. So instead, use selenium to navigate to the grafana ui and to click and get the link.

### CODE REVIEW

File: bin/export_grafana_snapshot.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
+import argparse
+import datetime
+import json
+import os
+import time
+
+import requests
+import requests.auth
```
Comment (mandarjog) [inline bin/export_grafana_snapshot.py]: pl Add a requirements file in the dir and add dependencies there.

Comment (mandarjog) [reply bin/export_grafana_snapshot.py]: Ensure you do autopep8 on the file, and add it to ci/cd

File: bin/export_grafana_snapshot.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    def export_snapshot(self, dashboard_name, external):
+        """
+        :param dashboard_name: dashboard to export snapshot
+        :param external: export locally or to raintank
+        """
+        dashboard_id = self.search_dashboard(dashboard_name)
+        dashboard =\
+            self.get_dashboard_by_uid(dashboard_id)
+        self.post_snapshot(dashboard,
```
Comment (mandarjog) [inline bin/export_grafana_snapshot.py]: Return value is ingnored.
We need to know the uuid where this snapshot is exported so we can associated tags.

What is the persistence story for local exports?

Comment (richardwxn) [reply bin/export_grafana_snapshot.py]: Ok, I guess you mean the url of where the snapshot exported? 

For local exports, it is persisted in local grafana db.

Do we still need another storage to store the exported snapshot information? Actually, there is api to query information of snapshots exported, also snapshot would be exported with name consist of dashboard name and timestamp or some other info.

File: bin/export_grafana_snapshot.py (hunk 1 above)
Comment (richardwxn) [reply bin/export_grafana_snapshot.py]: Added requirements file and did autopep8 check. Can you clarify how to add it to ci/cd

File: bin/export_grafana_snapshot.py (hunk 2 above)
Comment (mandarjog) [reply bin/export_grafana_snapshot.py]: We can maintain a grafana instance with a persistent store, but the grafana we will be exporting from may be destroyed. So we cannot just keep the data there.
We should associate is the istio build number with this data.

Comment (richardwxn) [reply bin/export_grafana_snapshot.py]: Please check the comment on top for update

**Comment by richardwxn** [comment]
> /lgtm
> 
> Can you specify what the limitation actually is?

I added comment above for description. Basically there is a gap between the get dashboard api and post snapshot api. If we hope to export the snapshot externally, the post snapshot api requires a dashboard object with panels data populated.  However the get dashboard api returns a dashboard without those. So in order to do this, we have to manually query datasource and do some manipulation. I tried to follow how the original grafana source code does but cannot guarantee fully recovered snapshot. And actually there is a feature request for a long time to make the snapshot api friendly to use but is still pending.

So I think it might be better to use selenium to make the job easier for now.

**Pull request merged**